### PR TITLE
File system mismatch when each pipeline uses separate file system.

### PR DESCRIPTION
**Pull request by mashhurs**
## Release notes

[rn:skip]

## What does this PR do?

Currently, Logstash is using `data/queue` path to persist the pipeline data and checks the file system size for required space when it runs. However, customers want such case that each pipeline will have its own file system, e.g:`data/queue/pipeline_id`. For this case, `data/queue` will not be an ideal because it maybe in different file system and Logstash checks different file system (`data/queue`) instead actual file system (`data/queue/pipeline_id`). This PR introduces to use pipeline file system size check instead `data/queue` file system.

## Why is it important/What is the impact to the user?

When users configure multiple persisted pipelines, they may face disk usage issue earlier than expected.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## How to test this PR locally

Configure multiple pipelines in `pipelines.yml` and run Logstash.
```
 - pipeline.id: test
   pipeline.workers: 1
   pipeline.batch.size: 1
   queue.type: persisted
   config.string: "input { stdin { } } output { stdout {} }"
 - pipeline.id: another_test
   queue.type: persisted
   config.string: "input { stdin { } } output { stdout {} }"
```

## Related issues

- Closes TBD

## Use cases

Example setup: `path.queue` is set to `/opt/logstash/queue` which is not in `sdl, sdk, sdj` file systems and each pipeline has its own file system, e.g: `/opt/logstash/queue/pipeline1` is in `/dev/sdl`.
```
$ ls -l /opt/logstash/queue
total 24
drwxrwsr-x. 3 root logstash 4096 Apr 12 11:48 pipeline1
drwxrwsr-x. 3 root logstash 4096 Apr 12 11:48 pipeline2
drwxrwsr-x. 3 root logstash 4096 Apr 12 11:48 pipeline3
$ df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay          60G   12G   46G  21% /
tmpfs            64M     0   64M   0% /dev
/dev/sda9        60G   12G   46G  21% /etc/hosts
shm              64M     0   64M   0% /dev/shm
tmpfs           8.0G   68K  8.0G   1% /opt/logstash/config/certs
/dev/sdl         49G   36K   49G   1% /opt/logstash/queue/pipeline1
/dev/sdk        2.0G   36K  1.9G   1% /opt/logstash/queue/pipeline2
/dev/sdj         49G   36K   49G   1% /opt/logstash/queue/pipeline3
tmpfs            32G     0   32G   0% /proc/acpi
tmpfs            32G     0   32G   0% /proc/scsi
tmpfs            32G     0   32G   0% /sys/firmware
$
```

## Logs

```
<u>$ ls data/queue 
another_test	test

<u>$ ls data/queue/test 
checkpoint.head	page.1

<u>$ ls data/queue/another_test 
checkpoint.head	page.0
```

### CODE REVIEW

File: config/logstash.yml
Comment (karenzone) [inline config/logstash.yml:218]:
```suggestion
# If queue.type: persisted, the directory path where the pipeline data files will be stored.  Logstash creates `pipeline.id` folder for each pipeline
# Default is path.data/queue.
```

Comment (karenzone) [reply config/logstash.yml:218]: Please consider moving the line about Logstash creating a folder for each pipeline to the first section.

Comment (karenzone) [inline config/logstash.yml:218]:
```suggestion
# If using queue.type: persisted, the directory path where the pipeline data files will be stored. Logstash creates `pipeline.id` folder for each pipeline.
# Default is path.data/queue.
```

**Review by karenzone — APPROVED** [review]
Left suggestions inline for your consideration.  Otherwise, LGTM.

**Review by jsvd — COMMENTED** [review]
We already support multiple pipelines with persistent queue enabled, which is written to `path.data/queue/pipeline_id`, so this PR doesn't seem to be changing that.

My understanding is that the original issue is a corner case where each individual `path.data/queue/<pipeline_id>` has its own filesystem (and is a symlink), but we use `path.data/queue` to validate the disk size.

Can we clarify exactly the issue and how this PR fixes it?

**Review by yaauie — COMMENTED** [review]
We have historically documented the PQ for a pipeline _as_ the directory that contains all of its interrelated files, not as those files themselves.

This PR solves an issue that is caused by users mounting additional filesystems inside their `path.queue` for each pipeline id, which relies on our internal implementation details and not our documented promises. While I am not opposed to solving a real in-the-wild issue, we already provide a way for each pipeline to provide its own `path.queue`, and therefore already provide a lever for a pipeline to have its queue files on a different filesystem.

I am wary of changes that advertise each pipeline's PQ directory as anything other than an atomic unit.

File: docs/static/persistent-queues.asciidoc
Comment (rpasche) [reply docs/static/persistent-queues.asciidoc:45]: To me, that *is* the directory, where `path.queue/pipeline_id` is pointing to. `path.queue` (for me) is more like the *parent* folder in which each *pipeline* will have its own directory, in which all pipeline related queue files are stored.

In our setup, we also could have set `path.queue` already to individual mounted filesystems, like
```
- pipeline.id: pipeline1
  queue.max_bytes: 49392123904
  path.queue: "/opt/logstash/queue/pipeline1"
- pipeline.id: pipeline2
  queue.type: "persisted"
  queue.max_bytes: 49392123904
  path.queue: "/opt/logstash/queue/pipeline2"
- pipeline.id: pipeline3
  queue.type: "persisted"
  queue.max_bytes: 1073741824
  path.queue: "/opt/logstash/queue/pipeline3"
```

But then, the directory containing the queue files for *pipeline1* would be `/opt/logstash/queue/pipeline1/pipeline1`, as the code at https://github.com/elastic/logstash/blob/8d01b1649dc848588357d27617c12584753de32f/logstash-core/src/main/java/org/logstash/ackedqueue/QueueFactoryExt.java#L56 is creating a directory named like `pipeline_id` below of `path.queue`.

This of course also works....it's just a minimal "visual" issue.

The PR that I initially created (https://github.com/elastic/logstash/pull/14211) and @mashhurs tuned into this one, should just make sure, that the *correct* filesystem is checked and used for calculating, if enough disk capacity is available.

This PQ validator is currently breaking our setup as we have multiple pipelines running in a single logstash instance....often 2 of those pipelines have queue sizes around 45 GB assigned. The "root" filesystem (where `/opt/logstash/queue` is stored) is "only" roughly the same size (45 GB). And that is just not enough to hold all configured queue capacity of all pipelines, using the current logic.

And yes....we could also configure only a single additional filesystem holding all queue data (and that be mounted at `/opt/logstash/queue`) but experiences in the past showed that we sometimes suffered from corrupted filesystems and in the end, we lost all data on those filesystems. Therefore, we decided to use individual disks for the different pipelines. In case a filesystem "breaks", we might only loose the queue events from a single pipeline, and not all.

Setting the `path.queue` (like in example above individually for each pipeline) should also fix the validator logic. But that would mean, that we would have to move all queue data (that might still exist now for pipelines) to a "subfolder" with the same name (for every pipeline in every single logstash instance in all of our clusters) so logstash version 8 is "finding" the queue data at the right place.

As `/opt/logstash/queue/pipeline1` is a mounted filesystem and `path.queue` is set to `/opt/logstash/queue`, this is how it looks like
```
$ ls -la /opt/logstash/queue/pipeline1
total 44
drwxrwsr-x. 4 root     logstash     4096 Jun  8 06:17 .
drwxr-xr-x. 1 logstash logstash     4096 Jun  6 13:13 ..
-rw-rw-r--. 1 logstash logstash        0 Mar 16 12:45 .lock
-rw-rw-r--. 1 logstash logstash        4 Mar 16 12:45 .queue-version
-rw-r--r--. 1 logstash logstash       34 Jun  6 13:14 checkpoint.head
drwxrws---. 2 root     logstash    16384 Mar 16 12:36 lost+found
-rw-rw-r--. 1 logstash logstash 67108864 Jun  3 13:42 page.1754
```

To fix our setup, we would need to make the structure look like this
```
$ ls -la /opt/logstash/queue/pipeline1
total 44
drwxrwsr-x. 4 root     logstash     4096 Jun  8 06:17 .
drwxr-xr-x. 1 logstash logstash     4096 Jun  6 13:13 ..
drwxr-sr-x. 2 logstash logstash     4096 Jun  8 06:17 pipeline1
drwxrws---. 2 root     logstash    16384 Mar 16 12:36 lost+found
$

$ ls -la /opt/logstash/queue/pipeline1/pipeline1
total 44
drwxrwsr-x. 4 root     logstash     4096 Jun  8 06:17 .
drwxr-xr-x. 1 logstash logstash     4096 Jun  6 13:13 ..
-rw-rw-r--. 1 logstash logstash        0 Mar 16 12:45 .lock
-rw-rw-r--. 1 logstash logstash        4 Mar 16 12:45 .queue-version
-rw-r--r--. 1 logstash logstash       34 Jun  6 13:14 checkpoint.head
-rw-rw-r--. 1 logstash logstash 67108864 Jun  3 13:42 page.1754
```
This is not what we want. It would be a huge manual effort to do this.

File: logstash-core/lib/logstash/persisted_queue_config_validator.rb
Comment (mashhurs) [reply logstash-core/lib/logstash/persisted_queue_config_validator.rb:50]: If you take a look at the [PR description user created](https://github.com/elastic/logstash/pull/14211), they created a file system for each pipeline and mounted to the `/opt/logstash/queue/pipeline_id`. My assumption is the base path `/opt/logstash/queue` in a reality in different file system. So, giving a room for each pipeline to use its own file system would be an ideal logic.
`queue_path` will include the pipeline ID and when [defining the file system, the logic](https://github.com/mashhurs/logstash/blob/d886c5206717cc62ee5a15a7f265142b28c64130/logstash-core/lib/logstash/persisted_queue_config_validator.rb#L95) retrieves the exact file system rather then the base path volume located.

I still couldn't reproduce the case since my env does not allow to create FS/mount volume (only 1 I was able to) but posted my question to clarify the base paths' file system. Let me know if I am missing anything else.

Thanks.

File: docs/static/persistent-queues.asciidoc
Comment (mashhurs) [reply docs/static/persistent-queues.asciidoc:45]: Thanks @rpasche for the details, I agree with your comments above that linking each FS to pipeline `path.queue` would be a workaround/solution for now.

I assume `/opt/logstash/queue` is in a different FS (not in `/dev/sdl` & `/dev/sdk`), can you plz confirm (basically `df .` should show)? I am trying to reproduce the case but my env's ability to create FSs and mounting Volumes is limited, so far no success. If you can pull my changes and run Logstash after building, would be a great help as well.

Thanks.

Comment (rpasche) [reply docs/static/persistent-queues.asciidoc:45]: > Thanks @rpasche for the details, I agree with your comments above that linking each FS to pipeline `path.queue` would be a workaround/solution for now.
> 
> I assume `/opt/logstash/queue` is in a different FS (not in `/dev/sdl` & `/dev/sdk`), can you plz confirm (basically `df .` should show)? I am trying to reproduce the case but my env's ability to create FSs and mounting Volumes is limited, so far no success. If you can pull my changes and run Logstash after building, would be a great help as well.
> 
> Thanks.

```
$ df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay          60G   12G   46G  21% /
tmpfs            64M     0   64M   0% /dev
/dev/sda9        60G   12G   46G  21% /etc/hosts
shm              64M     0   64M   0% /dev/shm
tmpfs           8.0G   68K  8.0G   1% /opt/logstash/config/certs
/dev/sdl         49G   36K   49G   1% /opt/logstash/queue/pipeline1
/dev/sdk        2.0G   36K  1.9G   1% /opt/logstash/queue/pipeline2
/dev/sdj         49G   36K   49G   1% /opt/logstash/queue/pipeline3
tmpfs            32G     0   32G   0% /proc/acpi
tmpfs            32G     0   32G   0% /proc/scsi
tmpfs            32G     0   32G   0% /sys/firmware
$
```
Yes, @mashhurs , `/opt/logstash/queue` is _not_ on either **sdl**, **sdk**, **sdj**. It's on **sda** (shown here with `overlay`) which is 60 GB in size. *if* the root disk would have been big enough (for example 200GB) we wouldn't have recognized this issue.

File: config/logstash.yml
Comment (mashhurs) [reply config/logstash.yml:217]: Agree with your point, thanks for the details.

File: logstash-core/lib/logstash/persisted_queue_config_validator.rb
Comment (mashhurs) [reply logstash-core/lib/logstash/persisted_queue_config_validator.rb]: To add more context, linking pipeline path to file system (tested and customer also mentioned below) uglifies the path where `ln` includes additional the `pipeline_id` (customer wants: `data/queue/pipeline_id`, after linking path becomes `data/queue/pipeline_id/pipeline_id`). 

So, IMO our current approach (LS to check pipeline file system) seems a best practise for this case.

**Comment by rpasche** [comment]
@mashhurs I have rebuilt the logstash image and deployed it in our environment. Logstash is now again able to start. Since this is currently a "running container", I might not be able to start dedicated "logastash" binary commands that you might need to get some proof.

Only proof I can give you so far is the contents of the pq_validator script like
```
(⎈ |cluster:default)➜  ~/git/logstash-deployments git:(pq_validator_test) k exec  -n datainput test-container -- cat /opt/logstash/logstash-core/lib/logstash/persisted_queue_config_validator.rb | grep queue_path
Defaulted container "test-container" out of: test-container, sysctl-tcp-keepalive (init)
      queue_path_file_system = Hash.new # (String: queue path, String: file system)
        queue_path = Paths.get(config.settings.get("path.queue"), pipeline_id).to_s
        pq_page_glob = ::File.join(queue_path, "page.*")
        create_dirs(queue_path)
        file_system = get_file_system(queue_path)
        queue_path_file_system[queue_path] = file_system
      check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
    def check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
        queue_path_file_system
          .select { |queue_path, file_system| !FsUtil.hasFreeSpace(Paths.get(queue_path), required_free_bytes.fetch(file_system, 0)) }
          .map { |queue_path, file_system| "Persistent queue path #{queue_path} is unable to allocate #{required_free_bytes.fetch(file_system, 0)} more bytes on top of its current usage." }
    def get_file_system(queue_path)
      fs = Files.getFileStore(Paths.get(queue_path))
    def create_dirs(queue_path)
      Files.createDirectories(Paths.get(queue_path))
```

**Comment by rpasche** [comment]
> @mashhurs I have rebuilt the logstash image and deployed it in our environment. Logstash is now again able to start. Since this is currently a "running container", I might not be able to start dedicated "logastash" binary commands that you might need to get some proof.
> 
> Only proof I can give you so far is the contents of the pq_validator script like
> 
> ```
> (⎈ |cluster:default)➜  ~/git/logstash-deployments git:(pq_validator_test) k exec  -n datainput test-container -- cat /opt/logstash/logstash-core/lib/logstash/persisted_queue_config_validator.rb | grep queue_path
> Defaulted container "test-container" out of: test-container, sysctl-tcp-keepalive (init)
>       queue_path_file_system = Hash.new # (String: queue path, String: file system)
>         queue_path = Paths.get(config.settings.get("path.queue"), pipeline_id).to_s
>         pq_page_glob = ::File.join(queue_path, "page.*")
>         create_dirs(queue_path)
>         file_system = get_file_system(queue_path)
>         queue_path_file_system[queue_path] = file_system
>       check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
>     def check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
>         queue_path_file_system
>           .select { |queue_path, file_system| !FsUtil.hasFreeSpace(Paths.get(queue_path), required_free_bytes.fetch(file_system, 0)) }
>           .map { |queue_path, file_system| "Persistent queue path #{queue_path} is unable to allocate #{required_free_bytes.fetch(file_system, 0)} more bytes on top of its current usage." }
>     def get_file_system(queue_path)
>       fs = Files.getFileStore(Paths.get(queue_path))
>     def create_dirs(queue_path)
>       Files.createDirectories(Paths.get(queue_path))
> ```

And....although I enabled the `debug` loglevel, I did not find a hint, that the PQ validator had been executed. So I cannot show any output here.

**Comment by mashhurs** [comment]
> @mashhurs I have rebuilt the logstash image and deployed it in our environment. Logstash is now again able to start. Since this is currently a "running container", I might not be able to start dedicated "logastash" binary commands that you might need to get some proof.
> 
> Only proof I can give you so far is the contents of the pq_validator script like
> 
> ```
> (⎈ |cluster:default)➜  ~/git/logstash-deployments git:(pq_validator_test) k exec  -n datainput test-container -- cat /opt/logstash/logstash-core/lib/logstash/persisted_queue_config_validator.rb | grep queue_path
> Defaulted container "test-container" out of: test-container, sysctl-tcp-keepalive (init)
>       queue_path_file_system = Hash.new # (String: queue path, String: file system)
>         queue_path = Paths.get(config.settings.get("path.queue"), pipeline_id).to_s
>         pq_page_glob = ::File.join(queue_path, "page.*")
>         create_dirs(queue_path)
>         file_system = get_file_system(queue_path)
>         queue_path_file_system[queue_path] = file_system
>       check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
>     def check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
>         queue_path_file_system
>           .select { |queue_path, file_system| !FsUtil.hasFreeSpace(Paths.get(queue_path), required_free_bytes.fetch(file_system, 0)) }
>           .map { |queue_path, file_system| "Persistent queue path #{queue_path} is unable to allocate #{required_free_bytes.fetch(file_system, 0)} more bytes on top of its current usage." }
>     def get_file_system(queue_path)
>       fs = Files.getFileStore(Paths.get(queue_path))
>     def create_dirs(queue_path)
>       Files.createDirectories(Paths.get(queue_path))
> ```

Great! Thanks @rpasche.
I have also reproduced the case and confirm Logstash free space calculation mismatches if base path and pipeline path are in different file systems.

```
// some debuggings practise logs 
Files system of /opt/logstash/queue : /dev/disk3s5
File system of /opt/logstash/queue/test : /dev/disk0s3
File system of /opt/logstash/queue/another_test : /dev/disk3s4

// Logstash is considering /dev/disk3s5 file system size instead /dev/disk0s3
```
So, it seems our current logic resolves the case and let's wait for review if someone has any ideas.
Thanks.

**Comment by jsvd** [comment]
@rpasche as this is worked on, have you tried setting a `path.queue` for each individual pipeline, like so:

`> cat config/pipelines.yml`

```yaml
- pipeline.id: test_1
  pipeline.workers: 1
  pipeline.batch.size: 1
  config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
  dead_letter_queue.enable: true
  queue.type: persisted
  path.queue: "/tmp/1"
- pipeline.id: test_2
  pipeline.workers: 1
  pipeline.batch.size: 1
  config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
  dead_letter_queue.enable: true
  queue.type: persisted
  path.queue: "/tmp/2"
```

**Comment by richardgilm** [comment]
@businessbean FYI. Do you know if  @rpasche tested this?

> @rpasche as this is worked on, have you tried setting a `path.queue` for each individual pipeline, like so:
> 
> `> cat config/pipelines.yml`
> 
> ```yaml
> - pipeline.id: test_1
>   pipeline.workers: 1
>   pipeline.batch.size: 1
>   config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
>   dead_letter_queue.enable: true
>   queue.type: persisted
>   path.queue: "/tmp/1"
> - pipeline.id: test_2
>   pipeline.workers: 1
>   pipeline.batch.size: 1
>   config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
>   dead_letter_queue.enable: true
>   queue.type: persisted
>   path.queue: "/tmp/2"
> ```

**Comment by jsvd** [comment]
> @rpasche as this is worked on, have you tried setting a path.queue for each individual pipeline, like so:
> @businessbean FYI. Do you know if @rpasche tested this?

To answer my own question, there was a comment by @rpasche about path.queue per pipeline a week ago here: https://github.com/elastic/logstash/pull/14212#discussion_r891963120. 

In summary, the main concern with adopting path.queue per pipeline is that it requires a different directory structure, and transitioning multiple PQs from a single path.queue to multiple path.queue outside of path.data is cumbersome.

**Comment by jsvd** [comment]
> @rpasche as this is worked on, have you tried setting a path.queue for each individual pipeline, like so:
> @businessbean FYI. Do you know if @rpasche tested this?

To answer my own question, there was a comment by @rpasche about path.queue per pipeline a week ago here: https://github.com/elastic/logstash/pull/14212#discussion_r891963120. 

In summary, the main concern with adopting path.queue per pipeline is that it requires a different directory structure, and transitioning multiple PQs from a single path.queue to multiple path.queue outside of path.data is cumbersome.

**Comment by rajmohand** [comment]
> > @rpasche as this is worked on, have you tried setting a path.queue for each individual pipeline, like so:
> > @businessbean FYI. Do you know if @rpasche tested this?
> 
> To answer my own question, there was a comment by @rpasche about path.queue per pipeline a week ago here: [#14212 (comment)](https://github.com/elastic/logstash/pull/14212#discussion_r891963120).
> 
> In summary, the main concern with adopting path.queue per pipeline is that it requires a different directory structure, and transitioning multiple PQs from a single path.queue to multiple path.queue outside of path.data is cumbersome.

Yes @richardgilm, @jsvd as mentioned by @rpasche we also tested the `path.queue ` it works for the validator logic but in the end it will leads to a different directory structure and lot of migration in production, potential dataloss.
with out the `path.queue` it was working fine in the 7.17.1.

File: logstash-core/lib/logstash/persisted_queue_config_validator.rb
Comment (kaisecheng) [inline logstash-core/lib/logstash/persisted_queue_config_validator.rb:52]: @mashhurs It works well with mounting the filesystem to the queue folder.
I have tired to use symlink pointing to path.queue/pipeline_id
It fails with with FileAlreadyExistsException in create_dirs() and [QueueFactory](https://github.com/elastic/logstash/blob/v7.17.0/logstash-core/src/main/java/org/logstash/ackedqueue/QueueFactoryExt.java#L60)

I think we can track the problem in a new issue if we consider symlinks is a valid scenario. WDYT @jsvd ?

Comment (mashhurs) [reply logstash-core/lib/logstash/persisted_queue_config_validator.rb:52]: Thanks @kaisecheng , your setting is also valid and as we took offline to test together and share thoughts, I have added exception handling logic.

**Review by kaisecheng — APPROVED** [review]
LGTM. I have tested `mount` and `ln -s` to path.queue/pipeline_id. Both can take their filesystem. Thank for also finishing the bonus point to handle symlink case 🎊

**Comment by kaisecheng** [comment]
@mashhurs please update the release notes session before merge

**Comment by mashhurs** [comment]
@kaisecheng as we took offline, release note isn't generated yet. I will include once it is ready. Thanks for letting me know.

**Comment by jsvd** [comment]
@logstashmachine backport 8.3

**Pull request merged**
